PWGCF/MultiparticleCorrelations/Core/MuPa-MemberFunctions.h

```diff
@@ -1024,43 +1229,65 @@ Bool_t EventCuts(aod::Collision const& collision, aod::Tracks const& tracks)
-void FillEventHistograms(
-  aod::Collision const& collision, aod::Tracks const& tracks, const Int_t rs,
-  const Int_t ba) // reco or sim, before or after event cuts
+template <typename T>
+void FillEventHistograms(aod::Collision const& collision, T const& tracks, eRecSim rs, eBeforeAfter ba)
 {
-  // Fill all event histograms.
+  // Fill all event histograms for reconstructed or simulated data.
+
+  // a) Common to both reconstructed and simulated;
+  // b) Only reconstructed;
+  // c) Only simulated.
 
   if (fVerbose) {
     LOGF(info, "\033[1;32m%s\033[0m", __PRETTY_FUNCTION__);
   }
 
+  // a) Common to both reconstructed and simulated;
+
+  cout << Form("ceh_a.fEventHistograms[eNumberOfEvents][%d][%d] = ", rs, ba) << ceh_a.fEventHistograms[eNumberOfEvents][rs][ba] << endl;
+
   ceh_a.fEventHistograms[eNumberOfEvents][rs][ba]->Fill(0.5);
-  ceh_a.fEventHistograms[eTotalMultiplicity][rs][ba]->Fill(tracks.size());
-  // SelectedTracks => filled directly in void process( ... )
-  // ceh_a.fEventHistograms[eCentrality][rs][ba]->Fill(  TBI  );  => TBI
-  // 20231007 not ready yet
-  ceh_a.fEventHistograms[eVertex_x][rs][ba]->Fill(collision.posX());
-  ceh_a.fEventHistograms[eVertex_y][rs][ba]->Fill(collision.posY());
-  ceh_a.fEventHistograms[eVertex_z][rs][ba]->Fill(collision.posZ());
-  ceh_a.fEventHistograms[eNContributors][rs][ba]->Fill(collision.numContrib());
-
-} // void FillEventHistograms(aod::Collision const& collision, aod::Tracks
-// const& tracks, const Int_t rs, const Int_t ba); // reco or sim, before or
-// after event cuts
+
+  switch (rs) {
+    // b) Only reconstructed:
+    case eRec:
+      ceh_a.fEventHistograms[eTotalMultiplicity][eRec][ba]->Fill(tracks.size());
+      // SelectedTracks => filled directly in void process( ... )
+      // ceh_a.fEventHistograms[eCentrality][eRec][ba]->Fill(  TBI  );  => TBI
+      // 20231007 not ready yet
+      ceh_a.fEventHistograms[eVertex_x][eRec][ba]->Fill(collision.posX());
+      ceh_a.fEventHistograms[eVertex_y][eRec][ba]->Fill(collision.posY());
+      ceh_a.fEventHistograms[eVertex_z][eRec][ba]->Fill(collision.posZ());
+      ceh_a.fEventHistograms[eNContributors][eRec][ba]->Fill(collision.numContrib());
+      break;
+
+    // c) Only simulated:
+    case eSim:
+      ceh_a.fEventHistograms[eTotalMultiplicity][eSim][ba]->Fill(tracks.size());
+      // SelectedTracks => filled directly in void process( ... )
+      // ceh_a.fEventHistograms[eCentrality][eSim][ba]->Fill(  TBI  );  => TBI
+      // 20231007 not ready yet
+      /*
+      ceh_a.fEventHistograms[eVertex_x][eSim][ba]->Fill(collision.posX());
+      ceh_a.fEventHistograms[eVertex_y][eSim][ba]->Fill(collision.posY());
+      ceh_a.fEventHistograms[eVertex_z][eSim][ba]->Fill(collision.posZ());
+      ceh_a.fEventHistograms[eNContributors][eSim][ba]->Fill(collision.numContrib());
+      */
+      break;
+  } // switch(rs) {
+
+} // template <typename T> void FillEventHistograms(...)
```

Comment:
But I need here eRec/eSim always to be enum, and of fixed type eRecSim. Which other type do you have in mind here for this particular function argument, when I could benefit from templates?

In addition, implementation with switch + fixed number of enums is safe, in the sense that this argument has to be either eRec or eSim, which then safely I can use as a corresponding array index (0 or 1).

I doubt there is a big performance penalty for simple operations like this, but in any case I will perform code profiling and optimization at some point later, when all other functionalitied are added to my task, and when I see where most of execution time is going. 

On the last comment: yes. I need to template also on the collision table, so that I can process pure MC.

Reply:
What about 
```
template <eRecSim rs, typename T1, typename T2>
void FillEventHistograms(T1 const& collision, T2 const& tracks, eBeforeAfter ba)
```
Then inside you function you use the construction
```
if constexpr (rs == eRec) {
  ..................
}

if constexpr (rs == eGen) {
  ..................
}
```
and then the compiler decides what to do and the conditional jumps are not in the generated code.

For calling the function then you use 
```
FillEventHistograms<eRec>(collision,tracks, eBefore);
```
for instance.

If you do it that way, you don't need any variable to store the process you are going to do.
Just template in the same way the whole chain. The initial call in the corresponding process function is done with the corresponding `enum` value and then that value is propagated along the whole chain and used when needed.
The compiler takes care of replicate when is needed and no additional jumps on `if` conditions are introduced.

I told you, templates are not for fashion :)

Reply:
OK, implemented now this way, both for FillEventHistograms and FillParticleHistograms